### PR TITLE
fix(AxelarAuthVerifier): use u128 for weights and threshold

### PR DESCRIPTION
**Pull request by re1ro**
[AXE-4337]
* [x] `AxelarAuthVerifier`: use `u128` for `weights` and `threshold`

[AXE-4337]: https://axelarnetwork.atlassian.net/browse/AXE-4337?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: contracts/axelar-auth-verifier/src/contract.rs [hunk 1]
```diff
@@ -238,7 +238,7 @@ impl AxelarAuthVerifier {
 
             let WeightedSigner { weight, .. } = signers.signers.get(signer_index).unwrap();
 
-            total_weight = total_weight.add(&weight);
+            total_weight += weight;
```
Comment (milapsheth) [inline contracts/axelar-auth-verifier/src/contract.rs:241]: this might overflow by default. rust has a checked_add method, take a look at that

File: contracts/axelar-auth-verifier/src/testutils.rs [hunk 2]
```diff
@@ -60,15 +60,15 @@ pub fn generate_signer_set(
         .into_iter()
         .map(|(_, pk_hash, w)| WeightedSigner {
             signer: BytesN::<32>::from_array(env, &pk_hash),
-            weight: U256::from_u32(env, w),
```
Comment (milapsheth) [inline contracts/axelar-auth-verifier/src/testutils.rs:63]: define weight above as `u128` instead, and use `sum::<u128>`

File: contracts/axelar-auth-verifier/src/test.rs [hunk 3]
```diff
@@ -358,17 +357,15 @@ fn rotate_signers_fail_low_total_weight() {
...
         .signer_set
         .signers
         .iter()
         .map(|WeightedSigner { weight, .. }| weight)
-        .reduce(|acc, weight| acc.add(&weight))
+        .reduce(|acc, weight| acc + weight)
         .expect("Empty signers");
 
-    let new_threshold = total_weight.add(&one);
+    let new_threshold = total_weight + 1;
```
Comment (milapsheth) [inline contracts/axelar-auth-verifier/src/test.rs:368]: add a test to catch overflow (when total weight > u128_max)

File: contracts/axelar-auth-verifier/src/contract.rs [hunk 4]
```diff
@@ -258,23 +258,22 @@ pub fn validate_signers(env: &Env, weighted_signers: &WeightedSigners) -> bool {
...
         }
 
-        if signer.weight == zero {
+        if signer.weight == 0 {
             return false;
         }
 
         previous_signer = signer.signer;
-        total_weight = total_weight.add(&signer.weight);
+        total_weight += signer.weight;
```
Comment (milapsheth) [inline contracts/axelar-auth-verifier/src/contract.rs:273]: checked_add is missing here

File: contracts/axelar-auth-verifier/src/test.rs [hunk 5]
```diff
@@ -303,11 +302,41 @@ fn rotate_signers_fail_zero_weight() {
...
 
     // get last signer and modify its weight to zero
     if let Some(mut last_signer) = new_signers.signer_set.signers.get(last_index) {
-        last_signer.weight = U256::from_u32(&env, 0);
+        last_signer.weight = 0u128;
+        new_signers.signer_set.signers.set(last_index, last_signer);
+    }
+
+    // should throw an error, last signer weight is zero
+    let res = client.try_rotate_signers(&new_signers.signer_set, &false);
```
Comment (milapsheth) [inline contracts/axelar-auth-verifier/src/test.rs:314]: 👍

**Review by milapsheth — APPROVED** [review]

File: contracts/axelar-auth-verifier/src/contract.rs (hunk 4 above)
Comment (re1ro) [reply contracts/axelar-auth-verifier/src/contract.rs]: Good catch

File: contracts/axelar-auth-verifier/src/contract.rs [hunk 6]
```diff
@@ -258,23 +258,22 @@ pub fn validate_signers(env: &Env, weighted_signers: &WeightedSigners) -> bool {
...
         }
 
-        if signer.weight == zero {
+        if signer.weight == 0 {
             return false;
         }
 
         previous_signer = signer.signer;
-        total_weight = total_weight.add(&signer.weight);
+        total_weight = total_weight.checked_add(signer.weight).unwrap();
```
Comment (milapsheth) [inline contracts/axelar-auth-verifier/src/contract.rs:273]: add test coverage for this

File: contracts/axelar-auth-verifier/src/test.rs [hunk 7]
```diff
@@ -224,6 +224,30 @@ fn fail_validate_proof_threshold_not_met() {
...
+    let last_index = signers.signer_set.signers.len() - 1;
+
+    // get last signer and modify its weight to zero
+    if let Some(mut last_signer) = signers.signer_set.signers.get(last_index) {
+        last_signer.weight = u128::MAX - 1;
+        signers.signer_set.signers.set(last_index, last_signer);
+    }
+
+    let msg_hash = generate_random_payload_and_hash(env);
+    let mut proof = generate_proof(env, msg_hash.clone(), signers);
```
Comment (milapsheth) [inline contracts/axelar-auth-verifier/src/test.rs:246]:
```suggestion
    let proof = generate_proof(env, msg_hash.clone(), signers);
```
as per lint warning

File: contracts/axelar-auth-verifier/src/test.rs [hunk 8]
```diff
@@ -224,6 +224,30 @@ fn fail_validate_proof_threshold_not_met() {
     // should panic, all signatures are valid but total weight is below threshold
     client.validate_proof(&msg_hash, &proof);
 }
+#[test]
+#[should_panic()]
```
Comment (milapsheth) [inline contracts/axelar-auth-verifier/src/test.rs:228]: add expected error message as lint suggests

**Review by milapsheth — APPROVED** [review]

File: contracts/axelar-auth-verifier/src/test.rs (hunk 8 above)
Comment (re1ro) [reply contracts/axelar-auth-verifier/src/test.rs]: fixed
